tests/benchmark/stateful/bloatnet/test_single_opcode.py

```diff
@@ -0,0 +1,177 @@
+"""
+abstract: BloatNet single-opcode benchmark cases for state-related operations.
+
+   These tests focus on individual EVM opcodes (SLOAD, SSTORE) to measure
+   their performance when accessing many storage slots across pre-deployed
+   contracts. Unlike multi-opcode tests, these isolate single operations
+   to benchmark specific state-handling bottlenecks.
+"""
+
+import pytest
+
+from ethereum_test_forks import Fork
+from ethereum_test_tools import (
+    Account,
+    Alloc,
+    Block,
+    BlockchainTestFiller,
+    Transaction,
+    While,
+)
+from ethereum_test_vm import Bytecode
+from ethereum_test_vm import Opcodes as Op
+
+REFERENCE_SPEC_GIT_PATH = "DUMMY/bloatnet.md"
+REFERENCE_SPEC_VERSION = "1.0"
+
+# ERC20 function selectors
+BALANCEOF_SELECTOR = 0x70A08231  # balanceOf(address)
+APPROVE_SELECTOR = 0x095EA7B3  # approve(address,uint256)
+ALLOWANCE_SELECTOR = 0xDD62ED3E  # allowance(address,address)
+
+
+# SLOAD BENCHMARK ARCHITECTURE:
+#
+#   [Pre-deployed ERC20 Contract] ──── Storage slots for balances
+#           │
+#           │  balanceOf(address) → SLOAD(keccak256(address || slot))
+#           │
+#   [Attack Contract] ──CALL──► ERC20.balanceOf(random_address)
+#           │
+#           └─► Loop(i=0 to N):
+#                 1. Generate random address from counter
+#                 2. CALL balanceOf(random_address) → forces cold SLOAD
+#                 3. Most addresses have zero balance → empty storage slots
+#
+# WHY IT STRESSES CLIENTS:
+#   - Each balanceOf() call forces a cold SLOAD on a likely-empty slot
+#   - Storage slot = keccak256(address || balances_slot)
+#   - Random addresses ensure maximum cache misses
+#   - Tests client's sparse storage handling efficiency
+
+
+@pytest.mark.valid_from("Prague")
+def test_sload_empty_erc20_balanceof(
+    blockchain_test: BlockchainTestFiller,
+    pre: Alloc,
+    fork: Fork,
+    gas_benchmark_value: int,
+    address_stubs,
```

Comment:
I didn't quite expect the fixture to be directly used by the test tbh, it's an interesting workaround!

A couple things:
- It will produce unexpected behavior to someone running `execute` and not knowing the inner workings of this tests because it will try to use all the stubs, including those that are meant to be used by other tests.
- The test will change its behavior depending on the stubs passed to the parameter in `execute`, which is not inherently a bad thing, but I think we should really think about it.

Reply:
Yes. I basically did that because a lot of the tests we will do will just be the same code, but changing the contract against which we run it.
Thus I thought it would not make sense to duplicate code all the time. And instead, I should just reuse the same code for any number of contracts that share interface.

So it's the stubs what actually determines how the contract behaves.

I spoke with @kamilchodola about this and he's also not sure if this way will be the best for his tool. Nevertheless, IMO it's the best in regards code to maintain and simplicity.